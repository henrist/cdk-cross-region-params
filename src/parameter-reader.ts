import * as cdk from "@aws-cdk/core"
import * as cr from "@aws-cdk/custom-resources"

export interface ParameterReaderProps {
  parameterName: string
  region: string
  /**
   * Value that must be updated to check if the parameter has a new value.
   *
   * @default Date.now().toString()
   */
  nonce?: string
}

function removeLeadingSlash(value: string): string {
  return value.slice(0, 1) == "/" ? value.slice(1) : value
}

/**
 * Get value of a SSM parameter dynamically during deployment
 * with support to read cross-region.
 */
export class ParameterReader extends cr.AwsCustomResource {
  constructor(scope: cdk.Construct, id: string, props: ParameterReaderProps) {
    super(scope, id, {
      onUpdate: {
        service: "SSM",
        action: "getParameter",
        parameters: {
          Name: props.parameterName,
        },
        region: props.region,
        // Update physical id to fetch the latest version.
        physicalResourceId: cr.PhysicalResourceId.of(props.parameterName),
      },
      policy: cr.AwsCustomResourcePolicy.fromSdkCalls({
        resources: [
          cdk.Arn.format(
            {
              service: "ssm",
              region: props.region,
              resource: "parameter",
              resourceName: removeLeadingSlash(props.parameterName),
            },
            cdk.Stack.of(scope),
          ),
        ],
      }),
      installLatestAwsSdk: false,
    })

    const r = this.node.findChild("Resource").node
      .defaultChild as cdk.CfnResource
    r.cfnOptions.metadata = r.cfnOptions.metadata || {}
    r.cfnOptions.metadata.Nonce = props.nonce ?? Date.now().toString()
  }

  get parameterValue(): string {
    return this.getResponseField("Parameter.Value")
  }
}
