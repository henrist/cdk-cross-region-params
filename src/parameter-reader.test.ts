import { CfnParameter, ParameterType } from "@aws-cdk/aws-ssm"
import { App, Stack } from "@aws-cdk/core"
import "jest-cdk-snapshot"
import { ParameterReader } from "./parameter-reader"

test("ssm-parameter-reader", () => {
  const app = new App()
  const stack1 = new Stack(app, "Stack1")
  const parameterName = "/my/param"

  new CfnParameter(stack1, "Param", {
    type: ParameterType.STRING,
    name: parameterName,
    value: "test",
  })

  const stack2 = new Stack(app, "Stack2")
  const reader = new ParameterReader(stack2, "ParamReader", {
    parameterName,
    region: "eu-west-1",
    nonce: "123",
  })

  expect(reader.parameterValue).toContain("Token")
  expect(stack2).toMatchCdkSnapshot({
    ignoreAssets: true,
  })
})
