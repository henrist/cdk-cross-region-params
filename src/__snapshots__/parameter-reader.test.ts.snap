// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssm-parameter-reader 1`] = `
Object {
  "Parameters": Any<Object>,
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ParamReaderCustomResourcePolicy6D6224AD": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParameter",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:eu-west-1:",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/my/param",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ParamReaderCustomResourcePolicy6D6224AD",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ParamReaderDE9769DB": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "ParamReaderCustomResourcePolicy6D6224AD",
      ],
      "Metadata": Object {
        "Nonce": "123",
      },
      "Properties": Object {
        "Create": "{\\"service\\":\\"SSM\\",\\"action\\":\\"getParameter\\",\\"parameters\\":{\\"Name\\":\\"/my/param\\"},\\"region\\":\\"eu-west-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/my/param\\"}}",
        "InstallLatestAwsSdk": false,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{\\"service\\":\\"SSM\\",\\"action\\":\\"getParameter\\",\\"parameters\\":{\\"Name\\":\\"/my/param\\"},\\"region\\":\\"eu-west-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/my/param\\"}}",
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
